src/Xrefcheck/Scanners/Markdown.hs

```diff
@@ -328,17 +338,71 @@ nodeExtractInfo input@(Node _ _ nSubs) = do
               link = if null url then rName else url
 
-          let (rLink, rAnchor) = case T.splitOn "#" link of
-                [t] -> (t, Nothing)
-                t : ts -> (t, Just $ T.intercalate "#" ts)
-                [] -> error "impossible"
+          let (rLink, rAnchor) = splitLink link
 
           let rInfo = referenceInfo rLink
 
           return $ FileInfoDiff
             (DList.singleton $ Reference {rName, rPos, rLink, rAnchor, rInfo})
             DList.empty
 
+        extractAnchor :: Text -> ExtractorM FileInfoDiff
+        extractAnchor text = do
+          let mName = do
+                tag <- safeHead $ parseTags text
+                attributes <- case tag of
+                  TagOpen a attrs | T.toLower a == "a" -> Just attrs
+                  _ -> Nothing
+                (_, name) <- find (\(field, _) -> T.toLower field `elem` ["name", "id"]) attributes
+                pure name
+
+          case mName of
+            Just aName -> do
+              let aType = HandAnchor
+                  aPos  = toPosition pos
+              return $ FileInfoDiff
+                mempty
+                (pure $ Anchor {aType, aName, aPos})
+
+            Nothing -> do
+              return mempty
+
+        extractHtmlReference ::  [Attribute Text] -> Maybe PosInfo -> DList.DList Reference
+        extractHtmlReference attrs tagPos = fromMaybe mempty do
+          link <- findAttributes ["href"] attrs
+          let (rLink, rAnchor) = splitLink link
+          pure . DList.singleton $ Reference "" rLink rAnchor (toPosition tagPos) (referenceInfo rLink)
+
+        splitLink :: Text -> (Text, Maybe Text)
+        splitLink link = case T.splitOn "#" link of
+          [t]    -> (t, Nothing)
+          t : ts -> (t, Just $ T.intercalate "#" ts)
+          []     -> error "impossible"
+
+        extractHtmlImage :: [Attribute Text] -> Maybe PosInfo -> DList.DList Reference
+        extractHtmlImage attrs tagPos = fromMaybe mempty do
+          link <- findAttributes ["src"] attrs
+          pure . DList.singleton $ Reference "" link Nothing (toPosition tagPos) (referenceInfo link)
+
+        extractHtmlLink :: Text -> ExtractorM FileInfoDiff
+        extractHtmlLink text =
+          case safeHead $ parseTags text of
+            Just (TagOpen tag attrs) | T.toLower tag == "img" ->
+              pure $ mempty & fidReferences .~ extractHtmlImage attrs pos
+            Just (TagOpen tag attrs) | T.toLower tag == "a" -> do
+              let  reference = extractHtmlReference attrs pos
+              case DList.toList reference of
+                [ref] -> do
+                  -- The `cmark-gfm` package parses the link tag as three separate nodes:
+                  -- `HTML_INLINE` with an opening tag, a `TEXT` with text in between,
+                  -- and `HTML_INLINE` with a closing tag. So we keep the extracted link in a state and
+                  -- try to get associated text in the next node.
```

Comment:
As we have been discussing in recent PRs, this code seems to be more difficult to understand with each new feature that requires to modify it. At least, comments and separating chunks of code to functions help in clarifying it for the moment 👍